src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.cpp

```diff
@@ -29,49 +29,131 @@
+  jint threads_in_evac = Atomic::load_acquire(&_bits);
+  while (true) {
+    jint newval = decrement
+      ? (threads_in_evac - 1) | OOM_MARKER_MASK
+      : threads_in_evac | OOM_MARKER_MASK;
+
+    jint other = Atomic::cmpxchg(&_bits, threads_in_evac, newval);
+    if (other == threads_in_evac) {
+      // Success: wait for other threads to get out of the protocol and return.
+      break;
+    } else {
+      // Failure: try again with updated new value.
+      threads_in_evac = other;
+    }
   }
-  // At this point we are sure that no threads can evacuate anything. Raise
-  // the thread-local oom_during_evac flag to indicate that any attempt
-  // to evacuate should simply return the forwarding pointer instead (which is safe now).
-  ShenandoahThreadLocalData::set_oom_during_evac(Thread::current(), true);
 }
 
-void ShenandoahEvacOOMHandler::register_thread(Thread* thr) {
-  jint threads_in_evac = Atomic::load_acquire(&_threads_in_evac);
+bool ShenandoahEvacOOMCounter::try_increment()
+{
+  jint threads_in_evac = Atomic::load_acquire(&_bits);
 
-  assert(!ShenandoahThreadLocalData::is_oom_during_evac(Thread::current()), "TL oom-during-evac must not be set");
   while (true) {
-    // Check for OOM.
-    // If offender has OOM_MARKER_MASK, then loop until no more threads in evac
+    // Cannot enter evacuation if OOM_MARKER_MASK is set.
     if ((threads_in_evac & OOM_MARKER_MASK) != 0) {
-      wait_for_no_evac_threads();
-      return;
+      return false;
     }
 
-    jint other = Atomic::cmpxchg(&_threads_in_evac, threads_in_evac, threads_in_evac + 1);
+    jint other = Atomic::cmpxchg(&_bits, threads_in_evac, threads_in_evac + 1);
     if (other == threads_in_evac) {
       // Success: caller may safely enter evacuation
-      return;
+      return true;
     } else {
       threads_in_evac = other;
     }
   }
 }
 
+ShenandoahEvacOOMHandler::ShenandoahEvacOOMHandler()
+  : _num_counters(calc_num_counters()) {
+
+  assert(_num_counters > 0, "sanity");
+  _threads_in_evac = NEW_C_HEAP_ARRAY(ShenandoahEvacOOMCounter, _num_counters, mtGC);
+  for (int i = 0; i < _num_counters; i++) {
+    new (&_threads_in_evac[i]) ShenandoahEvacOOMCounter;
+  }
+}
+
+int ShenandoahEvacOOMHandler::calc_num_counters() {
+  // Scale the number of counter buckets with the number of CPUs to
+  // minimise contention.  Also make sure the number is a power of two
+  // so we can map hash values to buckets with a simple mask.
+  const int nproc = os::active_processor_count();
+  const int clamped = MAX2(1, MIN2(nproc, 128));
+  return round_up_power_of_2(clamped);
+}
+
+uint64_t ShenandoahEvacOOMHandler::hash_pointer(const void *p) {
+  // Bit mixing function from MurmurHash3
+  uint64_t key = (uintptr_t)p;
+  key ^= (key >> 33);
+  key *= UINT64_C(0xff51afd7ed558ccd);
+  key ^= (key >> 33);
+  key *= UINT64_C(0xc4ceb9fe1a85ec53);
+  key ^= (key >> 33);
+  return key;
+}
+
+ShenandoahEvacOOMCounter *ShenandoahEvacOOMHandler::counter_for_thread(Thread* t) {
```

Comment:
Very minor nit: we ususally put the * next to the type, i.e. ShenandoahEvacOOMCounter*.